File: eval/control/export.ts
// The W3 control group — export a muro building as a naive JSON model carrying coordinates.
//
// **This is not muro's canonical JSON.** The canonical form is muro's definition of equivalence,
// not the shape an engineer who has never seen muro would write to hold the same building. What
// this emits is the latter: a shape that **stores every fact muro derives** — areas, wall
// endpoints, absolute opening positions, storey heights, the area of a grouping.
//
// Why export by machine rather than write one by hand. Four criteria, in the order they bite:
//   1. **It carries the same information.** If the control holds less, the outcome is about
//      missing information, not about derivation.
//   2. **A third party can re-run it.** A hand-written file cannot be regenerated; an exporter can.
//   3. **It cannot be dismissed as a strawman.** An objection to the shape must point at this file.
//   4. **Only the thing under test differs** — whether derivation exists.
//
// For (4) this export carries none of muro's relational machinery. A wall is not "the boundary of
// two spaces" but **a thing with endpoints and a thickness**; there are no default walls, so every
// wall is listed. Room identifiers keep muro's paths verbatim — mangling them would only handicap
// the control for no reason.

import { writeFileSync } from "node:fs";
import { derive } from "../../src/core/derive.js";
import type { Form, FormBoundary } from "../../src/core/derive.js";
import type { Attrs, Model } from "../../src/core/model.js";
import { zoneAreaM2 } from "../../src/core/model.js";
import { parseFile } from "../../src/parse-file.js";

/** A level. The storey height is a derived quantity, so it is stored. */
export interface JsonLevel {
  name: string;
  z: number;
  ceilingHeight?: number;
  slabThickness?: number;
  /** How high walls and columns stand (muro derives this) */
  storeyHeight?: number;
}

/**
 * A room.
 *
 * `pieces` lists the convex pieces of the floor; the floor is their union. muro derives this from
 * the written region; here it is stored — **so widening a room means editing several places.**
 * `areaM2` is the same: not derived but saved, which makes agreement the writer's responsibility.
 */
export interface JsonRoom {
  id: string;
  level?: string;
  /** The free label, when one was written. koyu reads no meaning from it. */
  type?: string;
  name?: string;
  pieces: Array<Array<[number, number]>>;
  areaM2?: number;
  floorZ?: number;
  ceilingZ?: number;
  indoor: boolean;
  attrs?: Record<string, string | number>;
}

/** A wall — a thing, not a relation. The rooms on either side are held as references. */
export interface JsonWall {
  id: string;
  level?: string;
  sides: [string, string];
  start: [number, number];
  end: [number, number];
  /** Thickness in mm. An `open` relation carries no material, so no thickness. */
  thickness?: number;
  /** True when the wall does not block air or light (a handrail, a railing) */
  air: boolean;
  baseZ?: number;
  topZ?: number;
  kind: string;
  attrs?: Record<string, string | number>;
}

/** An opening — held at absolute coordinates rather than as a position along its wall. */
export interface JsonOpening {
  id: string;
  wall: string;
  kind: "door" | "window";
  name?: string;
  center: [number, number];
  width: number;
  sillZ: number;
  headZ: number;
  leafThickness: number;
  sliding: boolean;
}

export interface JsonColumn {
  id: string;
  level: string;
  center: [number, number];
  width: number;
  depth: number;
  baseZ: number;
  topZ: number;
}

/**
 * A grouping (muro's `zone`).
 *
 * **The area is stored, not derived.** In muro `zoneAreaM2` counts its members every time; here it
 * is a fact on the page, so widening one room means fixing this number by hand or the model no
 * longer agrees with itself. Membership follows the identifier prefix (`/L5/A/ldk` belongs to
 * `/L5/A`) — the convention JSON practice actually uses.
 */
export interface JsonGroup {
  id: string;
  name?: string;
  /** Total indoor floor area in m2 (a stored derived quantity) */
  areaM2: number;
  attrs?: Record<string, string | number>;
}

export interface JsonBuilding {
  unit: "mm";
  levels: JsonLevel[];
  rooms: JsonRoom[];
  walls: JsonWall[];
  openings: JsonOpening[];
  columns: JsonColumn[];
  groups: JsonGroup[];
  site?: Array<{ id: string; polygon: Array<[number, number]>; areaM2: number }>;
}

const pt = (p: { x: number; y: number }): [number, number] => [p.x, p.y];

/**
 * Copy attributes across verbatim.
 *
 * **This is what keeps the information equal.** A value such as `floor:オーク` never reaches the
 * Form (the Form carries identity only), yet it is written in the source. Dropping it would make
 * the control lose for holding less — which proves nothing about derivation.
 */
function attrsOf(a: Attrs, lift: readonly string[] = []): Record<string, string | number> | undefined {
  const keys = Object.keys(a).sort().filter((k) => !lift.includes(k));
  if (keys.length === 0) return undefined;
  const out: Record<string, string | number> = {};
  for (const k of keys) out[k] = a[k] as string | number;
  return out;
}

/** Wall identifiers — JSON has no relational identity, so a running number is the only option */
const wallId = (i: number): string => `W${String(i + 1).padStart(3, "0")}`;

/**
 * Turn one `FormBoundary` into one wall.
 *
 * A muro boundary reaches the Form **per segment** (one relation can hold several segments), so
 * here there are as many walls as segments — exactly how a naive JSON model holds them.
 */
function wallOf(b: FormBoundary, i: number, model: Model): JsonWall {
  const decl = model.boundaries[b.boundary];
  return {
    id: wallId(i),
    ...(b.level === undefined ? {} : { level: b.level }),
    sides: [b.a, b.b],
    start: [b.segment.x1, b.segment.y1],
    end: [b.segment.x2, b.segment.y2],
    ...(b.material === undefined
      ? {}
      : { thickness: b.material.t, baseZ: b.material.z0, topZ: b.material.z1 }),
    kind: b.kind,
    air: b.air,
    ...(decl === undefined || attrsOf(decl.attrs) === undefined ? {} : { attrs: attrsOf(decl.attrs) }),
  };
}

export function exportBuilding(model: Model, form: Form): JsonBuilding {
  const walls = form.boundaries.map((b, i) => wallOf(b, i, model));
  // An opening points at the wall it sits on. Form openings carry the boundary index and the
  // segment, so that pair is the key.
  const wallIndexOf = new Map<string, number>();
  for (const [i, b] of form.boundaries.entries()) {
    wallIndexOf.set(`${b.boundary}|${b.segment.x1},${b.segment.y1},${b.segment.x2},${b.segment.y2}`, i);
  }
  const openings: JsonOpening[] = form.openings.map((o, k) => {
    const key = `${o.boundary}|${o.segment.x1},${o.segment.y1},${o.segment.x2},${o.segment.y2}`;
    const wi = wallIndexOf.get(key);
    if (wi === undefined) throw new Error(`the opening ${o.ref} sits on no exported wall`);
    return {
      id: `O${String(k + 1).padStart(3, "0")}`,
      wall: wallId(wi),
      kind: o.kind,
      ...(o.name === undefined ? {} : { name: o.name }),
      center: [o.cx, o.cy] as [number, number],
      width: o.w,
      sillZ: o.z0,
      headZ: o.z1,
      leafThickness: o.t,
      sliding: o.sliding,
    };
  });

  return {
    unit: "mm",
    levels: form.levels.map((l) => ({
      name: l.name,
      z: l.z,
      ...(l.h === undefined ? {} : { ceilingHeight: l.h }),
      ...(l.slab === undefined ? {} : { slabThickness: l.slab }),
      ...(l.pitch === undefined ? {} : { storeyHeight: l.pitch }),
    })),
    rooms: form.spaces.map((s) => ({
      id: s.path,
      ...(s.level === undefined ? {} : { level: s.level }),
      ...(s.type === undefined ? {} : { type: s.type }),
      pieces: s.outline.map((ring) => ring.map(pt)),
      ...(s.areaM2 === undefined ? {} : { areaM2: s.areaM2 }),
      ...(s.z0 === undefined ? {} : { floorZ: s.z0 }),
      ...(s.z1 === undefined ? {} : { ceilingZ: s.z1 }),
      indoor: s.indoor,
      ...(() => {
        const a = model.spaces.get(s.path)?.attrs;
        if (a === undefined) return {};
        const bag = attrsOf(a, ["name"]);
        return {
          ...(typeof a["name"] === "string" ? { name: a["name"] } : {}),
          ...(bag === undefined ? {} : { attrs: bag }),
        };
      })(),
    })),
    walls,
    openings,
    columns: form.columns.map((c, k) => ({
      id: `C${String(k + 1).padStart(3, "0")}`,
      level: c.level,
      center: [c.x, c.y] as [number, number],
      width: c.w,
      depth: c.d,
      baseZ: c.z0,
      topZ: c.z1,
    })),
    groups: [...model.zones.values()]
      .map((z) => {
        const bag = attrsOf(z.attrs, ["name"]);
        return {
          id: z.path,
          ...(typeof z.attrs["name"] === "string" ? { name: z.attrs["name"] } : {}),
          areaM2: zoneAreaM2(model, z.path),
          ...(bag === undefined ? {} : { attrs: bag }),
        };
      })
      .sort((p, q) => (p.id < q.id ? -1 : p.id > q.id ? 1 : 0)),
    ...(form.site.length === 0
      ? {}
      : { site: form.site.map((s) => ({ id: s.path, polygon: s.points.map(pt), areaM2: s.areaM2 })) }),
  };
}

/**
 * Fold coordinate pairs onto one line.
 *
 * `JSON.stringify(_, null, 2)` breaks `[0, 0]` across four lines. **Verbosity must not handicap the
 * control** — what is under test is propagation, not line count or character count. A hand-written
 * JSON folds them the same way.
 */
export function compact(json: string): string {
  return json.replace(/\[\s+(-?\d+(?:\.\d+)?),\s+(-?\d+(?:\.\d+)?)\s+\]/g, "[$1, $2]");
}

// ---- CLI ----

if (process.argv[1]?.endsWith("export.ts") === true) {
  const [entry, out] = process.argv.slice(2);
  if (entry === undefined) {
    console.error("usage: tsx eval/control/export.ts <entry.muro> [out.json]");
    process.exit(2);
  }
  const model = parseFile(entry);
  const json = compact(JSON.stringify(exportBuilding(model, derive(model)), null, 2)) + "\n";
  if (out === undefined) process.stdout.write(json);
  else {
    writeFileSync(out, json);
    console.error(`wrote ${out}`);
  }
}

File: export/ifc/koyu_ifc/export.py
"""Write one koyu building out as IFC4.

The input is the single JSON document `bin/koyu-form.mjs` produces. Every outline in it was
already raised by koyu's own constructors, so **this module holds no geometry rule at all** — it
decides which IFC entities exist and how they are related, and nothing about where a face is.

**The wall is built at full length and the opening is cut out of it.** koyu's `material.panels`
is already divided by the openings, so the panels alone would give correct geometry. But then no
`IfcOpeningElement` exists and **which wall hosts a door is nowhere in the file**. So the wall is
built whole, the opening box is removed by `IfcRelVoidsElement`, and the leaf is fitted by
`IfcRelFillsElement`. The panels become the check instead: the volume left after the boolean must
equal theirs, and the two derivations never met on the way in.

**There is no OwnerHistory.** IFC4 makes it optional, and koyu's position is that git holds the
history. Inventing a person, an organisation and a timestamp would also make the output different
on every run.
"""

from __future__ import annotations

import json
from collections import defaultdict
from pathlib import Path

import ifcopenshell

from . import georeference, properties
from .geometry import Builder
from .identity import global_id, space_identity, wall_identity

LEAF_THICKNESS_MM = 40.0

# koyu's device names, and what IFC calls them.
RUN_ENTITY = {
    "stair": "IfcStair",
    "ramp": "IfcRamp",
    "escalator": "IfcTransportElement",
    "lift": "IfcTransportElement",
}
RUN_TYPE = {
    ("stair", "straight"): "STRAIGHT_RUN_STAIR",
    ("stair", "return"): "HALF_TURN_STAIR",
    ("ramp", "straight"): "STRAIGHT_RUN_RAMP",
    ("ramp", "return"): "HALF_TURN_RAMP",
    ("escalator", "straight"): "ESCALATOR",
    ("escalator", "return"): "ESCALATOR",
    ("lift", "straight"): "ELEVATOR",
    ("lift", "return"): "ELEVATOR",
}


# The canonical JSON spellings this exporter knows how to read.
#
# **Read the version, or do not carry one.** koyu stamps `format` on every document precisely
# so a reader can refuse a spelling it does not know; this exporter took the document apart
# without ever looking, so a key rename would have produced a wrong IFC rather than an error.
# It did: `koyu-canonical/1.2` became `1.3` when the language version key was renamed, and
# nothing here noticed.
#
# A minor bump adds keys and leaves the ones already there alone, so reading a newer minor is
# safe and stays listed. A major bump renames or reorders, and is not.
READS_CANONICAL = ("koyu-canonical/1.2", "koyu-canonical/1.3")


def export(data: dict) -> ifcopenshell.file:
    canonical = data["canonical"]
    declared = canonical.get("format")
    if declared not in READS_CANONICAL:
        raise ValueError(
            f"This is koyu-ifc, which reads {' / '.join(READS_CANONICAL)}. "
            f"The document says {declared!r}. Nothing is wrong with the document — "
            "the exporter has not learnt this spelling of the canonical form."
        )
    name = canonical.get("name") or "koyu"
    space_attrs = {p: s.get("attrs", {}) for p, s in canonical.get("spaces", {}).items()}
    boundary_attrs = _boundary_attrs(canonical)

    f = ifcopenshell.file(schema="IFC4")

    # ---- Units. koyu is millimetres, so the numbers pass through unchanged ----
    units = f.create_entity(
        "IfcUnitAssignment",
        Units=[
            f.create_entity("IfcSIUnit", UnitType="LENGTHUNIT", Prefix="MILLI", Name="METRE"),
            f.create_entity("IfcSIUnit", UnitType="AREAUNIT", Name="SQUARE_METRE"),
            f.create_entity("IfcSIUnit", UnitType="VOLUMEUNIT", Name="CUBIC_METRE"),
            f.create_entity("IfcSIUnit", UnitType="PLANEANGLEUNIT", Name="RADIAN"),
        ],
    )

    # ---- The representation context. True north rides here ----
    context = f.create_entity(
        "IfcGeometricRepresentationContext",
        ContextType="Model",
        CoordinateSpaceDimension=3,
        Precision=1e-5,
        WorldCoordinateSystem=f.create_entity(
            "IfcAxis2Placement3D",
            Location=f.create_entity("IfcCartesianPoint", Coordinates=(0.0, 0.0, 0.0)),
        ),
        TrueNorth=_true_north(f, canonical.get("azimuth")),
    )
    body = f.create_entity(
        "IfcGeometricRepresentationSubContext",
        ContextIdentifier="Body",
        ContextType="Model",
        ParentContext=context,
        TargetView="MODEL_VIEW",
    )
    b = Builder(f, body)
    vocabulary = properties.Vocabulary(f)

    # ---- Georeferencing. Both halves of the frame or neither ----
    if "origin" in canonical and "azimuth" in canonical:
        georeference.write(f, context, canonical["origin"], float(canonical["azimuth"]))

    # ---- Spatial structure ----
    project = f.create_entity(
        "IfcProject",
        GlobalId=global_id("project", name),
        Name=name,
        UnitsInContext=units,
        RepresentationContexts=[context],
    )
    site = f.create_entity(
        "IfcSite",
        GlobalId=global_id("site", name),
        Name=data["site"][0]["path"] if data["site"] else "Site",
        ObjectPlacement=b.local_placement(),
        Representation=b.footprint(data["site"][0]["points"]) if data["site"] else None,
        CompositionType="ELEMENT",
        RefElevation=float(canonical["origin"]["elevation"])
        if "origin" in canonical and "elevation" in canonical["origin"]
        else None,
    )
    if data["site"]:
        properties.attach_quantities(
            f,
            site,
            f"site:{data['site'][0]['path']}",
            "Qto_SiteBaseQuantities",
            [properties.area(f, "GrossArea", data["site"][0]["areaM2"])],
        )
    building = f.create_entity(
        "IfcBuilding",
        GlobalId=global_id("building", name),
        Name=name,
        ObjectPlacement=b.local_placement(),
        CompositionType="ELEMENT",
    )
    _aggregate(f, project, [site], "project-site")
    _aggregate(f, site, [building], "site-building")

    storeys = {}
    for level in data["levels"]:
        storeys[level["name"]] = f.create_entity(
            "IfcBuildingStorey",
            GlobalId=global_id("storey", level["name"]),
            Name=level["name"],
            ObjectPlacement=b.local_placement(z=float(level["z"])),
            CompositionType="ELEMENT",
            Elevation=float(level["z"]),
        )
    _aggregate(f, building, list(storeys.values()), "building-storeys")

    # **Every product is placed against the building, not against its storey.**
    #
    # koyu's Form carries absolute coordinates — a space on the fifth floor already has its real z
    # — so hanging a product off a storey placement that also carries the storey's elevation adds
    # the elevation twice and lifts the whole floor. The storey keeps its own placement and its
    # `Elevation` attribute, which is what a tool reads to know where the storey datum is; which
    # storey a product belongs to is stated by `IfcRelContainedInSpatialStructure`, which is the
    # normative way to say it and does not depend on the placement chain.
    #
    # The alternative — making every coordinate relative to its storey — is the more usual IFC
    # shape, but koyu spaces may span storeys (an entrance hall through two floors), so there is
    # no one storey to be relative to.
    def placed_in(_level_name):
        return b.local_placement(building.ObjectPlacement)

    elements = defaultdict(list)

    # ---- Spaces ----
    spaces = {}
    by_storey_spaces = defaultdict(list)
    for s in data["spaces"]:
        attrs = space_attrs.get(s["path"], {})
        product = f.create_entity(
            "IfcSpace",
            GlobalId=global_id("space", space_identity(s["path"], attrs)),
            Name=s["path"],
            LongName=attrs.get("name"),
            ObjectPlacement=placed_in(s.get("level")),
            Representation=b.shape(
                b.extrude(piece, float(s.get("z0", 0.0)), float(s.get("z1", 0.0)))
                for piece in s["outline"]
            ),
            CompositionType="ELEMENT",
            PredefinedType="INTERNAL" if s.get("indoor") else "EXTERNAL",
        )
        spaces[s["path"]] = product
        by_storey_spaces[s.get("level")].append(product)
        properties.space(f, product, s, attrs)
        if isinstance(attrs.get("use"), str):
            vocabulary.classify(product, "use", attrs["use"])
    for level_name, members in by_storey_spaces.items():
        if level_name in storeys:
            _aggregate(f, storeys[level_name], members, f"storey-spaces-{level_name}")

    # ---- The outside. A space with no region never reaches the Form, so it comes from the
    # canonical form. IFC4 has a type for exactly this, and without it every envelope boundary
    # would have nothing on its far side ----
    external = {}
    for path, s in sorted(canonical.get("spaces", {}).items()):
        if s.get("attrs", {}).get("outside") != 1 or path in spaces:
            continue
        external[path] = f.create_entity(
            "IfcExternalSpatialElement",
            GlobalId=global_id("external", space_identity(path, s.get("attrs", {}))),
            Name=path,
            LongName=s.get("attrs", {}).get("name"),
            ObjectPlacement=b.local_placement(),
            PredefinedType="EXTERNAL",
        )
    if external:
        f.create_entity(
            "IfcRelReferencedInSpatialStructure",
            GlobalId=global_id("references", "external"),
            RelatingStructure=site,
            RelatedElements=list(external.values()),
        )

    # ---- Walls. Cut the openings, fit the leaves ----
    seen_ref = defaultdict(int)
    walls = {}
    element_of_wall = []
    for w in data["walls"]:
        index = seen_ref[w["ref"]]
        seen_ref[w["ref"]] += 1
        identity = wall_identity(w["ref"], index)
        w["identity"] = identity
        attrs = boundary_attrs.get(w["ref"], {})
        is_external = w["a"] not in spaces or w["b"] not in spaces
        if not w.get("material"):
            # A boundary of kind `open` has a line and no matter. IFC keeps a type for exactly
            # this, and `IfcRelSpaceBoundary.RelatedBuildingElement` is not optional, so the
            # relation cannot be written without one.
            virtual = f.create_entity(
                "IfcVirtualElement",
                GlobalId=global_id("virtual", identity),
                Name=identity,
                ObjectPlacement=placed_in(w.get("level")),
            )
            element_of_wall.append(virtual)
            elements[w.get("level")].append(virtual)
            continue
        material = w["material"]
        product = f.create_entity(
            "IfcWall",
            GlobalId=global_id("wall", identity),
            Name=identity,
            ObjectPlacement=placed_in(w.get("level")),
            Representation=b.shape(
                [b.extrude(material["outline"], float(material["z0"]), float(material["z1"]))]
            ),
            PredefinedType="STANDARD",
        )
        walls[(w["ref"], index)] = (product, w["segment"])
        element_of_wall.append(product)
        elements[w.get("level")].append(product)
        properties.wall(f, product, w, attrs, is_external)
        if isinstance(attrs.get("spec"), str):
            vocabulary.associate_material(product, attrs["spec"])

    # ---- Door and window types. koyu's `asset` is IFC's type object, so it maps across whole ----
    types = {}
    for asset_name, asset in sorted(canonical.get("assets", {}).items()):
        style = asset.get("attrs", {}).get("style")
        is_door = asset["kind"] == "door"
        # IFC4 requires an operation on a door type and a partitioning on a window type. koyu's
        # `style` is a coarse word (`hinged` / `sliding` / `auto`) and IFC's enumerations name a
        # hand as well — so rather than inventing "sliding to the left", the koyu word goes in
        # under `USERDEFINED`, which is what that value is for. Nothing is claimed that the source
        # did not say.
        kind_args = (
            {
                "OperationType": "USERDEFINED" if isinstance(style, str) else "NOTDEFINED",
                **({"ElementType": style} if isinstance(style, str) else {}),
            }
            if is_door
            else {"PartitioningType": "NOTDEFINED"}
        )
        types[asset_name] = f.create_entity(
            "IfcDoorType" if is_door else "IfcWindowType",
            GlobalId=global_id("type", asset_name),
            Name=asset_name,
            PredefinedType="NOTDEFINED",
            **kind_args,
        )
        properties.attach_pset(
            f,
            types[asset_name],
            f"type:{asset_name}",
            "koyu",
            [properties.single(f, f"koyu.{k}", v) for k, v in sorted(asset.get("attrs", {}).items())],
        )
    by_type = defaultdict(list)

    # ---- Openings ----
    opening_attrs = _opening_attrs(canonical)
    for op in data["openings"]:
        wall = _wall_for_opening(walls, op)
        if wall is None:
            continue
        void = f.create_entity(
            "IfcOpeningElement",
            GlobalId=global_id("opening", op["ref"]),
            Name=op["ref"],
            ObjectPlacement=b.local_placement(wall.ObjectPlacement),
            Representation=b.shape(
                [b.extrude(op["cutOutline"], float(op["z0"]), float(op["z1"]))]
            ),
            PredefinedType="OPENING",
        )
        f.create_entity(
            "IfcRelVoidsElement",
            GlobalId=global_id("voids", op["ref"]),
            RelatingBuildingElement=wall,
            RelatedOpeningElement=void,
        )
        leaf = _leaf_outline(op)
        filler = f.create_entity(
            "IfcDoor" if op["kind"] == "door" else "IfcWindow",
            GlobalId=global_id("filler", op["ref"]),
            Name=op.get("name") or op["ref"],
            ObjectPlacement=b.local_placement(wall.ObjectPlacement),
            Representation=b.shape([b.extrude(leaf, float(op["z0"]), float(op["z1"]))]),
            OverallHeight=float(op["z1"]) - float(op["z0"]),
            OverallWidth=float(op["w"]),
        )
        f.create_entity(
            "IfcRelFillsElement",
            GlobalId=global_id("fills", op["ref"]),
            RelatingOpeningElement=void,
            RelatedBuildingElement=filler,
        )
        elements[op.get("level")].append(filler)
        attrs = opening_attrs.get(op["ref"], {})
        properties.opening(f, filler, op, attrs)
        ref = attrs.get("ref")
        if isinstance(ref, str) and ref in types:
            by_type[ref].append(filler)
    for asset_name, members in by_type.items():
        f.create_entity(
            "IfcRelDefinesByType",
            GlobalId=global_id("typed", asset_name),
            RelatedObjects=members,
            RelatingType=types[asset_name],
        )

    # ---- Slabs, columns, the uncounted segments ----
    # A space whose region is not convex is derived as several pieces, and each piece carries its
    # own floor. So the identity counts within (space, kind) exactly as a wall counts within a ref.
    seen_slab = defaultdict(int)
    for slab in data["slabs"]:
        key = f"{slab['space']}/{slab['kind']}"
        index = seen_slab[key]
        seen_slab[key] += 1
        elements[slab.get("level")].append(
            f.create_entity(
                "IfcSlab",
                GlobalId=global_id("slab", f"{key}#{index}"),
                Name=f"{slab['space']} {slab['kind']}",
                ObjectPlacement=placed_in(slab.get("level")),
                Representation=b.shape(
                    [b.extrude(slab["outline"], float(slab["z0"]), float(slab["z1"]))]
                ),
                PredefinedType="ROOF" if slab["kind"] == "roof" else "FLOOR",
            )
        )

    for col in data["columns"]:
        product = f.create_entity(
            "IfcColumn",
            GlobalId=global_id("column", col["ref"]),
            Name=col["ref"],
            ObjectPlacement=placed_in(col.get("level")),
            Representation=b.shape(
                [b.extrude(col["outline"], float(col["z0"]), float(col["z1"]))]
            ),
        )
        elements[col.get("level")].append(product)
        if isinstance(col.get("attrs", {}).get("spec"), str):
            vocabulary.associate_material(product, col["attrs"]["spec"])

    # A `seg` is an uncounted stretch of a boundary that carries its own attributes — a change of
    # finish partway along a wall. IFC has no first-class notion of "part of a wall with different
    # properties", so it becomes a covering on that stretch, which is what it describes.
    for s in data["segs"]:
        wall_ref = s["ref"].rsplit("/", 1)[0]
        host = next((w for (r, _i), (w, _seg) in walls.items() if r == wall_ref), None)
        product = f.create_entity(
            "IfcCovering",
            GlobalId=global_id("seg", s["ref"]),
            Name=s["ref"],
            ObjectPlacement=placed_in(s.get("level")),
            Representation=b.shape([b.extrude(s["outline"], 0.0, float(s["t"]))]),
            PredefinedType="CLADDING",
        )
        elements[s.get("level")].append(product)
        if host is not None:
            f.create_entity(
                "IfcRelCoversBldgElements",
                GlobalId=global_id("covers", s["ref"]),
                RelatingBuildingElement=host,
                RelatedCoverings=[product],
            )

    # ---- Vertical circulation ----
    for r in data["runs"]:
        tag = f"{r['path']}@{r['level']}"
        entity = RUN_ENTITY[r["device"]]
        product = f.create_entity(
            entity,
            GlobalId=global_id("run", tag),
            Name=tag,
            ObjectPlacement=placed_in(r.get("level")),
            Representation=b.shape(
                b.prism(s["outline"], s["bottom"], s["top"]) for s in r["solids"]
            ),
            PredefinedType=RUN_TYPE[(r["device"], r["form"])],
        )
        elements[r.get("level")].append(product)
        properties.run(f, product, r)

    for level_name, members in elements.items():
        if level_name in storeys and members:
            f.create_entity(
                "IfcRelContainedInSpatialStructure",
                GlobalId=global_id("contains", f"storey:{level_name}"),
                RelatingStructure=storeys[level_name],
                RelatedElements=members,
            )

    # ---- Zones. koyu's counted aggregation is IFC's group of spaces ----
    for path, zone in sorted(canonical.get("zones", {}).items()):
        members = [p for q, p in spaces.items() if q == path or q.startswith(path + "/")]
        if not members:
            continue
        attrs = zone.get("attrs", {})
        group = f.create_entity(
            "IfcZone",
            GlobalId=global_id("zone", space_identity(path, attrs)),
            Name=path,
            LongName=attrs.get("name"),
        )
        f.create_entity(
            "IfcRelAssignsToGroup",
            GlobalId=global_id("assigns", path),
            RelatedObjects=members,
            RelatingGroup=group,
        )
        if isinstance(attrs.get("use"), str):
            vocabulary.classify(group, "use", attrs["use"])

    # ---- Space boundaries. **What a working export drops first, koyu holds as a first-class
    # edge**, so it is written rather than inferred ----
    for w, element in zip(data["walls"], element_of_wall):
        physical = "VIRTUAL" if element.is_a("IfcVirtualElement") else "PHYSICAL"
        for side, other in (("a", "b"), ("b", "a")):
            subject = spaces.get(w[side]) or external.get(w[side])
            if subject is None:
                continue
            outside = w[other] not in spaces
            f.create_entity(
                "IfcRelSpaceBoundary",
                # The identity carries the wall's index, because one `boundary` declaration can
                # be realised as several segments and each gets its own relation.
                GlobalId=global_id("boundary", f"{w['identity']}/{side}/{w[other]}"),
                Name=w["ref"],
                RelatingSpace=subject,
                RelatedBuildingElement=element,
                PhysicalOrVirtualBoundary=physical,
                InternalOrExternalBoundary="EXTERNAL" if outside else "INTERNAL",
            )

    return f


# ---- Small parts ----


def _true_north(f, azimuth):
    """True north in the model's own coordinates.

    `azimuth` is the true bearing of the +Y axis, clockwise from north, so true north is +Y turned
    counter-clockwise by that much. At 0 it is (0,1) = +Y; at 90 it is (-1,0) = -X.
    """
    if azimuth is None:
        return None
    import math

    a = math.radians(float(azimuth))
    return f.create_entity("IfcDirection", DirectionRatios=(-math.sin(a), math.cos(a)))


def _aggregate(f, whole, parts, tag: str):
    if not parts:
        return
    f.create_entity(
        "IfcRelAggregates",
        GlobalId=global_id("aggregates", tag),
        RelatingObject=whole,
        RelatedObjects=parts,
    )


def _canonical_ref(bd: dict, index: int) -> str:
    """The ref the Form gives a boundary, rebuilt from the canonical spelling.

    A boundary is identified by its two ends plus its place in **canonical** order — `a|b@i`,
    where `i` indexes the canonical boundaries array. Declaration order is a thing the canonical
    form discards, so indexing by it would give a different spelling for the same building. Both
    sides therefore agree without either asking the other's implementation.
    """
    return f"{bd['between'][0]}|{bd['between'][1]}@{index}"


def _boundary_attrs(canonical: dict) -> dict:
    """Boundary attributes keyed by the ref the Form uses."""
    return {
        _canonical_ref(bd, i): bd.get("attrs", {})
        for i, bd in enumerate(canonical.get("boundaries", []))
    }


def _opening_attrs(canonical: dict) -> dict:
    """Opening attributes keyed by ref, with the asset name kept under `ref`."""
    out = {}
    for i, bd in enumerate(canonical.get("boundaries", [])):
        ref = _canonical_ref(bd, i)
        for j, op in enumerate(bd.get("openings", [])):
            attrs = dict(op.get("attrs", {}))
            if isinstance(op.get("ref"), str):
                attrs["ref"] = op["ref"]
            out[f"{ref}/{j}"] = attrs
    return out


def _wall_for_opening(walls: dict, op: dict):
    """The wall an opening is hosted in.

    One `boundary` declaration splits into several segments, so the ref alone does not pick a
    wall. **An opening already knows which segment it sits on**, so the segment decides.
    """
    boundary_ref = op["ref"].rsplit("/", 1)[0]
    seg = op["segment"]
    key = (seg["x1"], seg["y1"], seg["x2"], seg["y2"])
    fallback = None
    for (ref, _index), (wall, wall_seg) in walls.items():
        if ref != boundary_ref:
            continue
        if (wall_seg["x1"], wall_seg["y1"], wall_seg["x2"], wall_seg["y2"]) == key:
            return wall
        fallback = fallback or wall
    return fallback


def _leaf_outline(op: dict) -> list:
    """The leaf, centred in the opening and as thin as a leaf rather than as thick as the wall."""
    outline = op["outline"]
    cx = sum(p[0] for p in outline) / len(outline)
    cy = sum(p[1] for p in outline) / len(outline)
    scale = LEAF_THICKNESS_MM / max(1.0, float(op["t"]))
    # Shrink across the wall only: the band's long axis is the opening's width, which stays.
    return [
        [cx + (p[0] - cx) * (1.0 if _along(op, 0) else scale),
         cy + (p[1] - cy) * (1.0 if _along(op, 1) else scale)]
        for p in outline
    ]


def _along(op: dict, axis: int) -> bool:
    """Whether the opening's width runs along x (axis 0) or y (axis 1)."""
    seg = op["segment"]
    dx, dy = abs(seg["x2"] - seg["x1"]), abs(seg["y2"] - seg["y1"])
    return (dx >= dy) if axis == 0 else (dy > dx)


def main(argv: list) -> int:
    if len(argv) < 3:
        print("Usage: python -m koyu_ifc.export <form.json> <out.ifc>")
        return 2
    data = json.loads(Path(argv[1]).read_text(encoding="utf-8"))
    f = export(data)
    f.write(argv[2])
    print(f"Wrote {argv[2]}")
    return 0


if __name__ == "__main__":
    import sys

    raise SystemExit(main(sys.argv))
